python/examples/core/harness.py

```diff
@@ -229,6 +229,56 @@ def _run_benchmark_n_iters(callback: Callable[[int], None], n_iters: int,
   return np.asarray([_pytimed(callback, *args) for _ in range(n_iters)])
 
 
+def _parse_problem_sizes(argument: str) -> Sequence[Union[int, Sequence[int]]]:
+  """Parse a problem size argument into a possibly nested integer sequence.
+
+  Examples:
+  64,128 -> [64, 128]
+  32,32,[1,1] -> [32, 32, [1, 1]]
+  """
+  problem_sizes = []
+  inc = []
+  dec = []
+  for idx, token in enumerate(argument.split(',')):
+    if token.startswith('['):
+      token = token[1:]
+      inc.append(idx)
+    if token.endswith(']'):
+      token = token[:-1]
+      dec.append(idx)
+    if not (len(inc) == len(dec) or len(inc) == len(dec) + 1):
+      raise ValueError()
+    problem_sizes.append(int(token))
+  # Nest the sizes found between the inc and dec indices.
+  if len(inc) != len(dec):
+    raise ValueError()
+  nested = []
+  curr = 0
+  for lb, ub in zip(inc, dec):
+    nested.extend(problem_sizes[curr:lb])
+    nested.append(problem_sizes[lb:ub + 1])
+    curr = ub + 1
+  nested.extend(problem_sizes[curr:])
+  return nested
+
+
+def _parse_read_only(argument: str) -> Sequence[str]:
+  """Parse a sequence of read only dimensions.
+
+  Examples:
+  k,m -> ['k', 'm']
+  [] -> []
+  """
+  if argument == '[]':
+    return []
+  read_only = []
+  for token in argument.split(','):
```

Comment:
same here: 
```
read_only = argument.split(')
for x in read_only:
  assert x.isalpha()
```

Also, is it "read_only" or "compile_time_symbols" ?
Can we use a better name?